Filter SNVs, INDELs, CNVs and SVs

fellen31 opened the pull request:
Closes #329, calls it filtered instead of clinical_set to be a bit more general. 

<!--
# genomic-medicine-sweden/nallo pull request

Many thanks for contributing to genomic-medicine-sweden/nallo!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/genomic-medicine-sweden/nallo/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/genomic-medicine-sweden/nallo/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core pipelines lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

fellen31 commented on CHANGELOG.md:
```diff
@@ -131,21 +135,22 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+| longphase             |             | 1.7.3       |
+| genmod                | 3.8.2       | 3.9         |
+| WhatsHap              | 2.2         | 2.3         |
+| SVDB                  |             | 2.8.1       |
```
```suggestion
| SVDB                  |             | 2.8.2       |
```

fellen31 commented on assets/schema_hgnc_ids.json:
```diff
@@ -0,0 +1,19 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "$id": "https://raw.githubusercontent.com/genomic-medicine-sweden/nallo/master/assets/schema_hgnc_ids.json",
+    "title": "genomic-medicine-sweden/nallo pipeline - params.hgnc_ids schema",
+    "description": "Schema for the file provided with params.hgnc_ids",
```
```suggestion
    "title": "genomic-medicine-sweden/nallo pipeline - params.filter_variants_hgnc_ids schema",
    "description": "Schema for the file provided with params.filter_variants_hgnc_ids",
```

Schmytzi reviewed (commented):
Good work but some things that should be adressed (mostly related to gawk)

fellen31 commented on workflows/nallo.nf:
```diff
@@ -101,9 +103,14 @@ workflow NALLO {
     ch_databases                = createReferenceChannelFromSamplesheet(params.snp_db, 'assets/schema_snp_db.json')
     ch_vep_plugin_files         = createReferenceChannelFromSamplesheet(params.vep_plugin_files, 'assets/schema_vep_plugin_files.json', Channel.value([]))
+    ch_hgnc_ids                 = createReferenceChannelFromSamplesheet(params.filter_variants_hgnc_ids, 'assets/schema_hgnc_ids.json')
+        .map { integer -> 'HGNC:' + integer[0].toString() } // Converts [ 256 ] to a string HGNC:256
```
It's because HGNC IDs need to be filtered like `1234` in hg19 but like `HGNC:1234` in hg38. Scout, which we will use in Stockholm to outputs gene panels export them without `HGNC:`.

It would be possible to accept the HGNC prefix into the pipeline if you think that makes more sense, then we will have to do some pre-processing. Might actually be more correct if `HGNC:` is in the actual ID. 
<img width="203" alt="Screenshot 2024-11-11 at 13 53 31" src="https://github.com/user-attachments/assets/64fb0f79-512e-4434-a59f-94442e5cb682">

fellen31 replied:
Actually, seems like VEP can use both now after 113 update. I will remove this and update the input schema.

fellen31 commented:
- Removed the `HGNC:` part in nallo.nf.
- Removed `#` from the header requirements and updated testdata (now `hgnc_id`). 
- Check for both integer and string types in schema to match both e.g. `4826` and `HGNC:4826`.

jemten commented on docs/output.md:
```diff
@@ -215,6 +215,21 @@ If the pipeline is run with phasing, the aligned reads will be happlotagged usin
+
+| Path                                             | Description                                   |
+| ------------------------------------------------ | --------------------------------------------- |
+| `snvs/{project}/{project}_*_filtered.vcf.gz`     | VCF file with filtered variants for a project |
+| `snvs/{project}/{project}_*_filtered.vcf.gz.tbi` | Index of the filtered VCF file                |
```
any reason for using `project` here and `family` for SVs?

fellen31 replied:
No, I will update to family here. It's a miss.

jemten commented on conf/modules/filter_variants.config:
```diff
@@ -0,0 +1,77 @@
+
+    withName: '.*:FILTER_VARIANTS_SVS:BCFTOOLS_VIEW' {
+        ext.prefix = {
+            def parts = []
+            parts << "${meta.id}"
+            parts << (params.skip_cnv_calling ? 'svs_merged' : 'svs_cnvs_merged')
+            if (!params.skip_sv_annotation) parts << 'annotated'
+            if (!params.skip_rank_variants) parts << 'ranked'
+            parts << 'filtered'
+            return parts.join('_')
```
Is it possible to use the function simpleName instead? Something like this 
`ext.prefix = {  ${input.simpleName}_filtered" }`

fellen31 replied:
It would be simpler.

For this workflow it would be possible since the filtering is done on the merged files (as opposed to chromosome regions), but it would break if the sample name contains dots. 

Perhaps we should restrict the input sample names to characters without dots if we want to use simpleName (same for raredisease).

jemten replied:
Yeah dots in sample names would break this. I think it would be worth looking into restricting dots in sample names in order to simplify the setting of prefixes. Not urgent though and in a future PR

jemten commented on nextflow_schema.json:
```diff
@@ -468,6 +468,24 @@
+                "filter_variants_hgnc_ids": {
+                    "type": "string",
+                    "pattern": "^\\S+\\.(csv|tsv)$",
+                    "description": "A tsv/csv file with a `#hgnc_ids` column header, and then one numerical HGNC ID per row. E.g. `4281`, not `HGNC:4281`.",
```
In the file specific schema it looked fine to have `HGNC:4281` as well as `4281`

Also shouldn't the header just be `hgnc_ids`? Alternatively update file schema.

fellen31 replied:
Either I missed changing the description, or messed up the merge. Will fix.

jemten commented on subworkflows/local/filter_variants/main.nf:
```diff
@@ -0,0 +1,37 @@
+
+    if ( filter_hgnc ) {
+
+
```
```suggestion
```

jemten commented on subworkflows/local/filter_variants/main.nf:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    BCFTOOLS_VIEW (
+        filter_hgnc ? ENSEMBLVEP_FILTERVEP.out.output.map { meta, vcf -> [ meta, vcf, [] ] } : ch_vcf.map { meta, vcf -> [ meta, vcf, [] ] },
```
I think that it would be clearer to set the output in the `filter_hgnc` block to overwrite `ch_vcf` so that the input for bcftools_view always is `ch_vcf.map { meta, vcf -> [ meta, vcf, [] ] }`

jemten commented on workflows/nallo.nf:
```diff
@@ -389,7 +398,9 @@ workflow NALLO {
 
         // Concat into family VCFs per family with all regions
-        BCFTOOLS_CONCAT ( ch_bcftools_concat_in )
+        BCFTOOLS_CONCAT (
+                ch_bcftools_concat_in
+            )
```
align

jemten commented on workflows/nallo.nf:
```diff
@@ -408,6 +419,22 @@ workflow NALLO {
+            //
+            // Filter SNVs
+            //
+            if(params.filter_variants_hgnc_ids || params.filter_snvs_expression != '') {
+
+                // Publish filtered `project` SNVs from here
+                FILTER_VARIANTS_SNVS (
+                    BCFTOOLS_SORT.out.vcf,
+                    ch_hgnc_ids,
+                    params.filter_variants_hgnc_ids
+                )
+                ch_versions = ch_versions.mix(FILTER_VARIANTS_SNVS.out.versions)
+
+            }
```
align

jemten commented on workflows/nallo.nf:
```diff
@@ -514,6 +541,24 @@ workflow NALLO {
     }
 
+        //
+        // Filter SVs
+        //
+        if(params.filter_variants_hgnc_ids || params.filter_svs_expression != '') {
+
+            if(params.skip_cnv_calling) {
+                ch_filter_svs_in = params.skip_sv_annotation ? CALL_SVS.out.family_vcf : params.skip_rank_variants ? ANN_CSQ_PLI_SVS.out.vcf : RANK_VARIANTS_SVS.out.vcf
+            } else {
+                ch_filter_svs_in = params.skip_sv_annotation ? annotate_svs_in : params.skip_rank_variants ? ANN_CSQ_PLI_SVS.out.vcf : RANK_VARIANTS_SVS.out.vcf
+            }
+
+            FILTER_VARIANTS_SVS (
+                ch_filter_svs_in,
+                ch_hgnc_ids,
+                params.filter_variants_hgnc_ids
+            )
+        }
```
align

jemten reviewed (commented):
Nice work! Had a couple of comments.

fellen31 commented on CHANGELOG.md:
```diff
@@ -140,21 +144,22 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+| samtools/merge        | 1.2         | 1.21        |
+| stranger              | 0.9.1       | 0.9.2       |
+| multiqc               | 1.21        | 1.25.1      |
+| ensemblvep/filter_vep |             | 113         |
```
Yes, but this is not possible at the moment with the SV problem Ram reported: https://github.com/Ensembl/ensembl-vep/issues/1774, and previous versions not working with hg38 HGNC IDs.

jemten reviewed (approved):
👍
